docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,479 @@
+
+## Solution
+As stated earlier, we would like to improve our current locality aware load balancing, and make it more flexible for users.
+### Use cases
+After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:
+
+1. Keeping traffic in single zone (disable all cross-zone traffic)
+2. Load balancing based on priorities within a zone. Example: node → AZ → region
+3. Prefer traffic in single zone with fallback to other zones
+4. Load balance equally to list of zones
+5. When leaving zone, target specific zones first. Example zones physically close to each other
+6. Load balance equally to local dpp and other zone
+7. Exclude zone from load balancing
+8. Load balance equally within a group of zones
+
+### API design
+Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
+Because of this we have decided that we will split our API in two sections: `localZone ` and `crossZone`.
+
+Here is the full example of policy:
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      disabled: false
+      localZone:
+        affinityTags:                                       # (1)
+        - key: "k8s.io/node"                                # (2)
+          weight: 90                                        # (3)
+        - key: k8s.io/az"
+      crossZone:                                            # (4)
+        failover:                                           # (5)
+          - from:                                           # (6)
+              zones: ["zone-1", "zone-2"]
+            to:                                             # (7)
+              type: Only                                    # (8)
+              zones: ["zone-1", "zone-2"]
+          - from:
+              zones: ["zone-3"]
+            to:
+              type: AnyExcept
+              zones: ["zone-1"]
+          - to:
+              type: Any
+      failoverThreshold: 
+        percentage: 70                                      # (7)
+```
+
+
+(1) In `localZone` section, you configure the affinity of tags used for load balancing. If list is empty or localZone is not specified, load balancing works normally, without taking any priorities into account. 
+
+(2) In `key` field, you define the name of the tag.
+
+(3) In `weight` field, you define proportional load that the group should take. Default is calculated based on the order and might and first in the list is going to have the highest weight while the last the lowest.
+
+Example:
+```yaml
+affinityTags:
+  - key: "k8s.io/node"
+  - key: k8s.io/az"
+  - key: k8s.io/region"
+```
+
+Would create weights:
+
+```yaml
+affinityTags:
```

Comment:
If I use `weight` does the order become irrelevant? For example, are these 2 equal?
```yaml
affinityTags:
  - key: "k8s.io/node"
    weight: 9
  - key: k8s.io/az"
    weight: 90
  - key: k8s.io/region"
    weight: 900
---
affinityTags:
  - key: k8s.io/region"
  - key: k8s.io/az"
  - key: "k8s.io/node"
```

Reply:
yes, when weight is configured order becomes irrelevant